### PR TITLE
Stable test: relaxation

### PR DESCRIPTION
**Pull request by andrea-petrocchi**

### CODE REVIEW

File: tests/stable/test_relaxation.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+import mammos_entity as me
+import numpy as np
+import pytest
+
+
+def _write_p2_file(filename, theta, phi):
+    mx = np.sin(theta) * np.cos(phi)
+    my = np.sin(theta) * np.sin(phi)
+    mz = np.cos(phi)
```
Comment (lang-m) [inline tests/stable/test_relaxation.py]: Shouldn't this be

```suggestion
    mz = np.cos(theta)
```

File: tests/stable/test_relaxation.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            my = {my}
+            mz = {mz}
+
+            [field]
+            hx = 0
+            hy = 0
+            hz = 1
+            hstart = 0.0
+            hfinal = 0.0
+            hstep = -0.5
```
Comment (lang-m) [inline tests/stable/test_relaxation.py]: Why a negative number for the step?

File: tests/stable/test_relaxation.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        _write_krn_file(tmp_path / system_name, Js)
+
+        # run hysteresis loop without demag
+        cmd = shlex.split(f"{loop_bin} {system_name} --verbose")
+        res = subprocess.run(cmd, cwd=tmp_path)
+        res.check_returncode()
+
+        # extract Bc from loop
+        hystloop = me.from_csv(tmp_path / f"hyst_{system_name}" / "mammos_hysteresis.csv")
+        with subtests.test(msg=f"phi_deg={phi_deg}"):
```
Comment (lang-m) [inline tests/stable/test_relaxation.py]: Do we really need subtest when we already use parametrize?

File: tests/stable/test_relaxation.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            assert np.all(hystloop.J_par_T.value >= 0.9 * Js)
```
Comment (lang-m) [inline tests/stable/test_relaxation.py]: 90% alignment seems fairly general. What actual accuracy do we reach with the current implementation?

File: tests/stable/test_relaxation.py (hunk 2 above)
Comment (andrea-petrocchi) [reply tests/stable/test_relaxation.py]: I thought it was necessary, but I realized we can just remove it.

File: tests/stable/test_relaxation.py (hunk 3 above)
Comment (andrea-petrocchi) [reply tests/stable/test_relaxation.py]: 5 random angles phi are generated for each theta. In this way they are created randomly for each theta, I thought it was more general. You would put both theta and phi in the parametrize? Only one list of phi for all thetas? Probably it does not change much.

File: tests/stable/test_relaxation.py (hunk 4 above)
Comment (andrea-petrocchi) [reply tests/stable/test_relaxation.py]: Raised accuracy to 99%. Honestly accuracy seems to be 100% in all tests.

File: tests/stable/test_relaxation.py (hunk 3 above)
Comment (andrea-petrocchi) [reply tests/stable/test_relaxation.py]: I have parametrized phi as well. It does look more elegant.

File: tests/stable/test_relaxation.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    _write_krn_file(tmp_path / system_name, Js)
+
+    # run hysteresis loop without demag
+    cmd = shlex.split(f"{loop_bin} {system_name} --verbose")
+    res = subprocess.run(cmd, cwd=tmp_path)
+    res.check_returncode()
+
+    # extract Bc from loop
+    hystloop = me.from_csv(tmp_path / f"hyst_{system_name}" / "mammos_hysteresis.csv")
+    assert np.all(hystloop.J_par_T.value >= 0.99 * Js)
```
Comment (AdamsMP92) [inline tests/stable/test_relaxation.py]: We could additionally check here that the value does not exceed `Js` beyond machine precision. Otherwise, the test still accepts polarization values larger than `Js`.

File: tests/stable/test_relaxation.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            # theta (rad) phi (rad) K1 (J/m3) not used Js (Tesla) A (J/m)
+            0.0 0.0 {K1} 0.0 {Js} {A}
+            """
+        )
+    )
+
+
+@pytest.mark.parametrize("theta_deg", [15, 30, 45, 60, 75])
+@pytest.mark.parametrize("phi_deg", np.random.randint(0, 359, 5))
+def test_stoner_wohlfarth_switching(loop_bin, mesh_bin, tmp_path, subtests, theta_deg, phi_deg):
```
Comment (AdamsMP92) [inline tests/stable/test_relaxation.py]: Here it might be better to use the function name `test_stoner_wohlfarth_zero_field_relaxation`, as this test is not regarding switching. Is this right?

File: tests/stable/test_relaxation.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+            f"""\
+            # theta (rad) phi (rad) K1 (J/m3) not used Js (Tesla) A (J/m)
+            0.0 0.0 {K1} 0.0 {Js} {A}
+            """
+        )
+    )
+
+
+@pytest.mark.parametrize("theta_deg", [15, 30, 45, 60, 75])
+@pytest.mark.parametrize("phi_deg", np.random.randint(0, 359, 5))
```
Comment (AdamsMP92) [inline tests/stable/test_relaxation.py]: It would be more reproducible if we use fixed angles, which are not randomly generated, e.g., `@pytest.mark.parametrize("phi_deg", [0, 37, 90, 173, 271])`

File: tests/stable/test_relaxation.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        dedent(
+            f"""\
+            # theta (rad) phi (rad) K1 (J/m3) not used Js (Tesla) A (J/m)
+            0.0 0.0 {K1} 0.0 {Js} {A}
+            """
+        )
+    )
+
+
+@pytest.mark.parametrize("theta_deg", [15, 30, 45, 60, 75])
```
Comment (AdamsMP92) [inline tests/stable/test_relaxation.py]: For the theta angles it would be interesting to also include edge cases, i.e., `[1, 15, 45, 75, 89]` and it could be interesting to additionally use angles on the southern hemisphere, i.e., `[91, 105, 135, 165, 179]`. For the latter one, the magnetization should relax parallel to `-z`, while for the upper hemisphere to `+z`.

One could use something like: 
`@pytest.mark.parametrize(
    "theta_deg, expected_sign",
    [
        (1, +1),
        (15, +1),
        (45, +1),
        (75, +1),
        (89, +1),
        (91, -1),
        (105, -1),
        (135, -1),
        (165, -1),
        (179, -1),
    ],
)`

and then check the sign as well. And then: 

`expected_J_par = expected_sign * Js`
`np.testing.assert_allclose(
    hystloop.J_par_T.value,
    expected_J_par,
    rtol=1e-2,
    atol=0.0,
)`

**Review by AdamsMP92 — CHANGES_REQUESTED** [review]

File: tests/stable/test_relaxation.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+    # generate mesh
+    L = 20.0
+    h = 4.0  # Coarse for speed
+    cmd = shlex.split(f"{mesh_bin} --geom box --extent {L},{L},{L} --h {h} --out-name {system_name}")
+    res = subprocess.run(cmd, cwd=tmp_path)
+    res.check_returncode()
+
+    # write input files
+    Js = 1.6  # Tesla
```
Comment (AdamsMP92) [inline tests/stable/test_relaxation.py]: could you a entity here.

File: tests/stable/test_relaxation.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+def test_stoner_wohlfarth_zero_field_relaxation(
+    loop_bin, mesh_bin, tmp_path, subtests, theta_deg, phi_deg, expected_sign
+):
+    """Test switch in Stoner-Wohlfarth model."""
+    system_name = f"sw_{theta_deg}_{phi_deg}"
+    theta = np.deg2rad(theta_deg)
+    phi = np.deg2rad(phi_deg)
+
+    # generate mesh
+    L = 20.0
```
Comment (AdamsMP92) [inline tests/stable/test_relaxation.py]: could use an entity here.

**Review by AdamsMP92 — CHANGES_REQUESTED** [review]

**Review by AdamsMP92 — APPROVED** [review]
